File: LeetCode/212WordSearchII/WordSearchII.cpp
// TIME LIMIT EXCEEDED FOR CODE

class TrieNode {
public:
    map<char,TrieNode*> children;
    bool endOfWord = false;
    void addWord(string word) {
        TrieNode *curr = this;
        for (int i = 0; i < word.size(); i++) {
            if (curr->children.find(word[i]) == curr->children.end()) {
                curr->children[word[i]] = new TrieNode();
            }
            curr = curr->children[word[i]];
        }
        curr->endOfWord = true;
    }
};

class Solution {
public:
    void dfs(vector<vector<char>>& board, set<string> &ans_set, set<vector<int>> &visited, int r, int c, TrieNode* node, string word) {
        vector<int> char_index{r, c};
        if (r < 0 || c < 0 || r == board.size() || c == board[0].size()
            || visited.find(char_index) != visited.end()
            || node->children.find(board[r][c]) == node->children.end()) {
                return;
            }
        visited.insert(char_index);
        node = node->children[board[r][c]];
        word += board[r][c];
        if (node->endOfWord) {
            ans_set.insert(word);
        }
        dfs(board, ans_set, visited, r + 1, c, node, word);
        dfs(board, ans_set, visited, r - 1, c, node, word);
        dfs(board, ans_set, visited, r, c + 1, node, word);
        dfs(board, ans_set, visited, r, c - 1, node, word);
        visited.erase(char_index);
    }
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        // Trie + DFS -> Time:O(mn*4^(mn)), Space:O(4^(mn))
        vector<string> ans;
        TrieNode *root = new TrieNode();
        for (int i = 0; i < words.size(); i++) {
            root->addWord(words[i]);
        }

        int rows = board.size(), cols = board[0].size();
        set<string> ans_set;
        set<vector<int>> visited;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                dfs(board, ans_set, visited, i, j, root, "");
            }
        }

        for (string i:ans_set) {
            ans.push_back(i);
        }

        return ans;
    }
};

File: LeetCode/212WordSearchII/WordSearchII.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.endOfWord = False
    
    def add_word(self, word):
        curr = self
        for c in word:
            if c not in curr.children:
                curr.children[c] = TrieNode()
            curr = curr.children[c]
        curr.endOfWord = True

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        # Trie + DFS -> Time:O(mn*4^(mn)), Space:O(4^(mn))
        root = TrieNode()
        for w in words:
            root.add_word(w)
        
        ROWS, COLS = len(board), len(board[0])
        ans, visited = set(), set()

        def dfs(r, c, node, word):
            if (r < 0 or c < 0 or r == ROWS or c == COLS
                or (r, c) in visited
                or board[r][c] not in node.children):
                return
            visited.add((r, c))
            node = node.children[board[r][c]]
            word += board[r][c]
            if node.endOfWord:
                ans.add(word)
            dfs(r + 1, c, node, word)
            dfs(r - 1, c, node, word)
            dfs(r, c + 1, node, word)
            dfs(r, c - 1, node, word)
            visited.remove((r, c))

        for i in range(ROWS):
            for j in range(COLS):
                dfs(i, j, root, "")

        return list(ans)
